src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -571,97 +599,189 @@ private boolean createFeedbackForDuplicateTests(Result result, ProgrammingExerci
+        result.getFeedbacks().stream().filter(fb -> FeedbackType.AUTOMATIC.equals(fb.getType()) && fb.getText().equalsIgnoreCase(testCase.getTestName())).findFirst()
+                .ifPresent(feedback -> feedback.setCredits(credits));
+    }
+
+    /**
+     * Calculates the points that should be awarded for a successful test case.
+     * @param result used to determine if the calculation is performed for the solution.
+     * @param programmingExercise the result belongs to.
+     * @param test for which the points should be calculated.
+     * @param totalTestCaseCount in the given exercise.
+     * @param weightSum of all test cases in the exercise.
+     * @return the points which should be awarded for successfully completing the test case.
+     */
+    private double calculatePointsForSuccessfulTestCase(final Result result, final ProgrammingExercise programmingExercise, final ProgrammingExerciseTestCase test,
+            int totalTestCaseCount, double weightSum) {
+        final boolean isWeightSumZero = Precision.equals(weightSum, 0, 1E-8);
+        final double testPoints;
+
+        // a non-zero score has to be calculated for the solution even if the weight sum is zero
+        // to avoid a warning to the instructor that the solution has faults
+        // => each test case receives a weight of one
+        if (isWeightSumZero && result.getParticipation() instanceof SolutionProgrammingExerciseParticipation) {
+            testPoints = (1.0 / totalTestCaseCount) * programmingExercise.getMaxPoints();
+        }
+        else if (isWeightSumZero) {
+            // this test case must have zero weight as well; avoid division by zero
+            testPoints = 0D;
+        }
+        else {
+            double testWeight = test.getWeight() * test.getBonusMultiplier();
+            testPoints = (testWeight / weightSum) * programmingExercise.getMaxPoints();
+        }
+
+        return testPoints + test.getBonusPoints();
+    }
+
+    /**
+     * Calculates a total penalty that should be applied to the score.
+     *
+     * This includes the penalties from static code analysis and of submission policies.
+     *
+     * @param programmingExercise the participation belongs to.
+     * @param participation for which should be checked for possible penalties.
+     * @param staticCodeAnalysisFeedback automatic feedback from static code analysis.
+     * @param applySubmissionPolicy determines if the submission policy should be applied.
+     * @return a total penalty that should be deducted from the score.
+     */
+    private double calculateTotalPenalty(final ProgrammingExercise programmingExercise, final Participation participation, final List<Feedback> staticCodeAnalysisFeedback,
+            boolean applySubmissionPolicy) {
+        double penalty = 0;
+
+        if (Boolean.TRUE.equals(programmingExercise.isStaticCodeAnalysisEnabled()) && Optional.ofNullable(programmingExercise.getMaxStaticCodeAnalysisPenalty()).orElse(1) > 0) {
```

Comment:
Nothing you changed but does it make sense to set the default value to 1? Seems arbitrary. 0 would make more sense in my opinion. What do you think?

Reply:
At least some value >0 is needed, as a maximum penalty of `null` means that there is no maximum and even all points can be deducted.
I replaced it by 100 as that would be the maximum possible value if set.